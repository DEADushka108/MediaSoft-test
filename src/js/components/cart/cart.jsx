import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import Header from '../header/header';
import Footer from '../footer/footer';
import { setCartList } from '../../store/ships/selectors';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { AppRoute } from '../../utils/const';
import { ActionCreator } from '../../store/ships/ships';
import { redirectToRoute } from '../../store/redirect/redirect-action';

const Cart = (props) => {
  const {shipsInCart, onCloseClick, redirect} = props;
  const price = shipsInCart.reduce(function(prev, curr) { return prev + curr.price }, 0);
  return <Fragment>
    <Header/>
    <main className="page-main">
      <h1 className="visually-hidden">Cart</h1>
      <section className="cart">
        <ul className="cart__list">
          {shipsInCart.map((ship, index) => {
            return <li key={`${ship.name}-${index}`} className="cart__item">
              <button type="button" className="cart__close-button" onClick={() => {
                onCloseClick(ship);
              }}>
                <svg className="cart__close-icon">
                  <use xlinkHref="#cart-remove-item"></use>
                </svg>
              </button>
              <Link to={`${AppRoute.CARD}/${ship.id}`} className="cart__link">
                <img className="cart__image" src={ship.previewImage} alt={ship.name}></img>
                <p className="cart__text">{ship.name}</p>
                <p className="cart__text">{ship.price}$</p>
              </Link>
            </li>
          })}
        </ul>
        <div className="cart__total">
          <p className="cart__total-text">Total price:</p>
          <p className="cart__total-text">{price}$</p>
        </div>
        <div className="cart__button-group">
          <button className="cart__button-submit" type="button" disabled={shipsInCart.length === 0} onClick={() => {
            redirect(`${AppRoute.FORM}`);
          }}>Buy</button>
          <Link to={`${AppRoute.ROOT}`} className="cart__button-continue">Continue</Link> 
        </div>
      </section>
    </main>

    <Footer/>

  </Fragment>
};

const mapStateToProps = (state) => ({
  shipsInCart: setCartList(state),
})

const maoDispatchToProps = (dispatch) => ({
  onCloseClick(ship) {
    ship.isInCart = !ship.isInCart;
    dispatch(ActionCreator.updateShipStatus(ship));
  },
  redirect(route) {
    dispatch(redirectToRoute(route));
  },
})

export {Cart}
export default connect(mapStateToProps, maoDispatchToProps)(Cart);
