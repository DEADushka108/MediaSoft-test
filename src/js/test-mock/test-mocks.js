export * from './const/const';
export * from './ships/ships';
export * from './store/store';
