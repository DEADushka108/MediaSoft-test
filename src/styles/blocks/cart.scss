.cart {
  width: 320px;
  margin: 0 auto;
}

.cart__list {
  list-style: none;
  padding: 0;
  margin: 0;
  display: flex;
  justify-content: center;
  flex-direction: column;
}

.cart__item {
  display: flex;
  justify-content: space-between;
  padding-top: 40px;
  padding-bottom: 40px;
  border-top: 1px solid rgba($text-blue, 0.3);
  border-bottom: 1px solid rgba($text-blue, 0.3);
}

.cart__link {
  flex-basis: 90%;
  display: flex;
  justify-content: space-between;
  text-decoration: none;
  color: $text-light-blue;

  &:hover {
    color: rgba($text-blue, 0.5);
  }

  &:active {
    color: rgba($text-blue, 0.3);
  }
}

.cart__image {
  width: 60px;
  height: 60px;
}

.cart__text {
  width: 60px;
  align-self: center;
  text-align: end;
}

.cart__close-button {
  border: none;
  background-color: transparent;
}

.cart__close-icon {
  width: 10px;
  height: 10px;
  fill: $text-blue;
}

.cart__total {
  display: flex;
  justify-content: space-between;
}

.cart__total-text {
  font-size: 20px;
}

.cart__button-submit {
  margin-bottom: 20px;
  margin-left: 20px;
  align-self: center;
  width: 281px;
  text-decoration: none;
  display: inline-block;
  vertical-align: middle;
  box-sizing: border-box;
  padding: 11px 12px 14px 12px;
  border-radius: 10px;
  text-align: center;
  color: $text-white;
  font-weight: 700;
  background-color: $button-blue;
  border: none;
  cursor: pointer;
  box-shadow: 5px 10px 20px rgba(53, 110, 173, 0.2);

  &:hover,
  &:focus {
    color: $text-blue;
    background-color: transparent;
    padding: 8px 9px 11px 9px;
    border: 3px solid $text-blue;
  }

  &:active {
    color: rgba($text-blue, 0.3);
  }

  &:disabled {
    opacity: 0.2;
  }
}

.cart__button-continue {
  width: 281px;
  margin-bottom: 20px;
  margin-left: 20px;
  align-self: center;
  text-decoration: none;
  display: inline-block;
  vertical-align: middle;
  box-sizing: border-box;
  padding: 11px 12px 14px 12px;
  border-radius: 10px;
  text-align: center;
  color: $text-light-blue;
  font-weight: 700;
  background-color: transparent;
  border: none;
  cursor: pointer;

  &:hover,
  &:focus {
    color: $text-blue;
    background-color: transparent;
    box-shadow: 5px 10px 20px rgba(53, 110, 173, 0.2);
  }

  &:active {
    color: rgba($text-blue, 0.3);
  }
}

.cart__button-group {
  display: flex;
  flex-direction: column;
}

@media (min-width: $tablet-width) {
  .cart {
    width: 640px;
  }

  .cart__image {
    width: 100px;
    height: 100px;
  }

  .cart__text {
    width: 200px;
    font-size: 20px;
  }

  .cart__close-icon {
    width: 15px;
    height: 15px;
  }
}

@media (min-width: $desktop-width) {
  .cart {
    width: 1300px;
  }

  .cart__image {
    width: 320px;
    height: 240px;
  }

  .cart__text {
    width: 400px;
    font-size: 30px;
  }

  .cart__close-icon {
    width: 25px;
    height: 25px;
  }

  .cart__total-text {
    font-size: 30px;
  }

  .cart__button-submit,
  .cart__button-continue {
    font-size: 26px;
    width: 460px;
    text-transform: uppercase;
  }

  .cart__button-continue {
    text-transform: none;
  }
}
