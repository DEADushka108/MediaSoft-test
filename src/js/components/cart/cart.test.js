import React from 'react';
import {Provider} from 'react-redux';
import {MemoryRouter} from 'react-router-dom';
import renderer from 'react-test-renderer';
import {noop, shipsList, store} from '../../test-mock/test-mocks';
import {Cart} from './cart';

it(`Should Cart component render correctly`, () => {
  const tree = renderer
    .create((
      <Provider store={store}>
        <MemoryRouter>
          <Cart
            shipsInCart={shipsList}
            redirect={noop}
            onCloseClick={noop}
          />
        </MemoryRouter>
      </Provider>
    ), {
      createNodeMock() {
        return {};
      }
    })
    .toJSON();

  expect(tree).toMatchSnapshot();
});
