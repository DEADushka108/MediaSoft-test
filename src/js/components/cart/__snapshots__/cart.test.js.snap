// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Cart component render correctly 1`] = `
Array [
  <header
    className="page-header"
  >
    <div
      className="page-header__wrapper"
    >
      <a
        className="page-header__logo"
        href="/"
        onClick={[Function]}
      >
        Starshipsshop
      </a>
      <nav
        className="main-nav"
      >
        <ul
          className="main-nav__list"
        >
          <li
            className="main-nav__item"
          >
            <a
              className="main-nav__link main-nav__link--active"
              href="/"
              onClick={[Function]}
            >
              <svg
                className="main-nav__icon"
              >
                <use
                  xlinkHref="#home-page"
                />
              </svg>
              <span
                className="main-nav__text"
              >
                Shop
              </span>
            </a>
          </li>
          <li
            className="main-nav__item"
          >
            <a
              className="main-nav__link"
              href="/contact"
              onClick={[Function]}
            >
              <svg
                className="main-nav__icon"
              >
                <use
                  xlinkHref="#social-group"
                />
              </svg>
              <span
                className="main-nav__text"
              >
                Contact
              </span>
            </a>
          </li>
          <li
            className="main-nav__item"
          >
            <a
              className="main-nav__link"
              href="/cart"
              onClick={[Function]}
            >
              <svg
                className="main-nav__icon"
              >
                <use
                  xlinkHref="#shopping-cart"
                />
              </svg>
              <span
                className="main-nav__text"
              >
                Cart
              </span>
              <span
                className="main-nav__sub"
              >
                2
              </span>
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </header>,
  <main
    className="page-main"
  >
    <h1
      className="visually-hidden"
    >
      Cart
    </h1>
    <section
      className="cart"
    >
      <ul
        className="cart__list"
      >
        <li
          className="cart__item"
        >
          <button
            className="cart__close-button"
            onClick={[Function]}
            type="button"
          >
            <svg
              className="cart__close-icon"
            >
              <use
                xlinkHref="#cart-remove-item"
              />
            </svg>
          </button>
          <a
            className="cart__link"
            href="/card/3"
            onClick={[Function]}
          >
            <img
              alt="Battlestar Galactica"
              className="cart__image"
              src="./img/content/battlestar-galactica-preview.png"
            />
            <p
              className="cart__text"
            >
              Battlestar Galactica
            </p>
            <p
              className="cart__text"
            >
              234122
              $
            </p>
          </a>
        </li>
        <li
          className="cart__item"
        >
          <button
            className="cart__close-button"
            onClick={[Function]}
            type="button"
          >
            <svg
              className="cart__close-icon"
            >
              <use
                xlinkHref="#cart-remove-item"
              />
            </svg>
          </button>
          <a
            className="cart__link"
            href="/card/4"
            onClick={[Function]}
          >
            <img
              alt="Borg's Cube"
              className="cart__image"
              src="./img/content/borgs-cube-preview.jpg"
            />
            <p
              className="cart__text"
            >
              Borg's Cube
            </p>
            <p
              className="cart__text"
            >
              1002000
              $
            </p>
          </a>
        </li>
        <li
          className="cart__item"
        >
          <button
            className="cart__close-button"
            onClick={[Function]}
            type="button"
          >
            <svg
              className="cart__close-icon"
            >
              <use
                xlinkHref="#cart-remove-item"
              />
            </svg>
          </button>
          <a
            className="cart__link"
            href="/card/5"
            onClick={[Function]}
          >
            <img
              alt="Death Star"
              className="cart__image"
              src="./img/content/death-star-preview.png"
            />
            <p
              className="cart__text"
            >
              Death Star
            </p>
            <p
              className="cart__text"
            >
              1000000
              $
            </p>
          </a>
        </li>
        <li
          className="cart__item"
        >
          <button
            className="cart__close-button"
            onClick={[Function]}
            type="button"
          >
            <svg
              className="cart__close-icon"
            >
              <use
                xlinkHref="#cart-remove-item"
              />
            </svg>
          </button>
          <a
            className="cart__link"
            href="/card/6"
            onClick={[Function]}
          >
            <img
              alt="Eagle-5"
              className="cart__image"
              src="./img/content/eagle-5-preview.jpg"
            />
            <p
              className="cart__text"
            >
              Eagle-5
            </p>
            <p
              className="cart__text"
            >
              100
              $
            </p>
          </a>
        </li>
        <li
          className="cart__item"
        >
          <button
            className="cart__close-button"
            onClick={[Function]}
            type="button"
          >
            <svg
              className="cart__close-icon"
            >
              <use
                xlinkHref="#cart-remove-item"
              />
            </svg>
          </button>
          <a
            className="cart__link"
            href="/card/7"
            onClick={[Function]}
          >
            <img
              alt="Flying Dish"
              className="cart__image"
              src="./img/content/flying-dishes-preview.jpg"
            />
            <p
              className="cart__text"
            >
              Flying Dish
            </p>
            <p
              className="cart__text"
            >
              3400000
              $
            </p>
          </a>
        </li>
        <li
          className="cart__item"
        >
          <button
            className="cart__close-button"
            onClick={[Function]}
            type="button"
          >
            <svg
              className="cart__close-icon"
            >
              <use
                xlinkHref="#cart-remove-item"
              />
            </svg>
          </button>
          <a
            className="cart__link"
            href="/card/8"
            onClick={[Function]}
          >
            <img
              alt="Heart of Gold"
              className="cart__image"
              src="./img/content/golden-heart-preview.png"
            />
            <p
              className="cart__text"
            >
              Heart of Gold
            </p>
            <p
              className="cart__text"
            >
              100000000
              $
            </p>
          </a>
        </li>
        <li
          className="cart__item"
        >
          <button
            className="cart__close-button"
            onClick={[Function]}
            type="button"
          >
            <svg
              className="cart__close-icon"
            >
              <use
                xlinkHref="#cart-remove-item"
              />
            </svg>
          </button>
          <a
            className="cart__link"
            href="/card/9"
            onClick={[Function]}
          >
            <img
              alt="Millenium Falcon"
              className="cart__image"
              src="./img/content/millennium-falcon-preview.png"
            />
            <p
              className="cart__text"
            >
              Millenium Falcon
            </p>
            <p
              className="cart__text"
            >
              54980
              $
            </p>
          </a>
        </li>
      </ul>
      <div
        className="cart__total"
      >
        <p
          className="cart__total-text"
        >
          Total price:
        </p>
        <p
          className="cart__total-text"
        >
          105691202
          $
        </p>
      </div>
      <div
        className="cart__button-group"
      >
        <button
          className="cart__button-submit"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Buy
        </button>
        <a
          className="cart__button-continue"
          href="/"
          onClick={[Function]}
        >
          Continue
        </a>
      </div>
    </section>
  </main>,
  <footer
    className="page-footer"
  >
    <div
      className="page-footer__wrapper"
    >
      <a
        className="page-footer__logo"
        href="/"
        onClick={[Function]}
      >
        Starshipsshop
      </a>
      <div
        className="page-footer__copyright"
      >
        <p
          className="page-footer__text"
        >
          Â© 2020 Stars ships shop Ltd.
        </p>
      </div>
    </div>
  </footer>,
]
`;
